src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -0,0 +1,248 @@
+                    }
+                }
+
+                return _reachableLabels;
+            }
+        }
+
+        /// <summary>
+        /// Return a list of all the nodes reachable from this one, in a topologically sorted order.
+        /// </summary>
+        public ImmutableArray<BoundDecisionDag> TopologicallySortedNodes()
+        {
+            // We use an iterative topological sort to avoid overflowing the compiler's runtime stack for a large switch statement.
+            return TopologicalSort.IterativeSort<BoundDecisionDag>(new[] { this }, successorFunction);
+
+            IEnumerable<BoundDecisionDag> successorFunction(BoundDecisionDag dag)
+            {
+                switch (dag)
+                {
+                    case BoundEvaluationPoint p:
+                        yield return p.Next;
+                        yield break;
+                    case BoundDecisionPoint p:
+                        yield return p.WhenTrue;
+                        yield return p.WhenFalse;
+                        yield break;
+                    case BoundWhenClause p:
+                        yield return p.WhenTrue;
+                        if (p.WhenFalse != null)
+                        {
+                            yield return p.WhenFalse;
+                        }
+                        yield break;
+                    case BoundDecision p:
+                        yield break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(dag);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Given a decision dag and a constant-valued input, produce a simplified decision dag that has removed all the
+        /// tests that are unnecessary due to that constant value. This simplification affects flow analysis (reachability
+        /// and definite assignment) and permits us to simplify the generated code.
+        /// </summary>
+        internal BoundDecisionDag SimplifyDecisionDagForConstantInput(
+            BoundExpression input,
+            Conversions conversions,
+            DiagnosticBag diagnostics)
+        {
+            ConstantValue inputConstant = input.ConstantValue;
+            Debug.Assert(inputConstant != null);
+
+            // First, we topologically sort the nodes of the dag so that we can translate the nodes bottom-up.
+            // This will avoid overflowing the compiler's runtime stack which would occur for a large switch
+            // statement if we were using a recursive strategy.
+            ImmutableArray<BoundDecisionDag> sortedNodes = this.TopologicallySortedNodes();
+
+            // Cache simplified/translated replacement for each translated dag node
+            var replacementCache = PooledDictionary<BoundDecisionDag, BoundDecisionDag>.GetInstance();
+
+            // Loop backwards through the topologically sorted nodes to translate them, so that we always visit a node after its successors
```

Comment:
>Loop backwards [](start = 15, length = 14)

Should `TopologicalSort` return the nodes, leaves first instead?

Reply:
Slightly useful to the clients but I think that would be nonstandard enough to be confusing. For example it wouldn't exactly be a topological sort. I could rename successors to predecessors, but that would be really strange given the direction of the edges in the dag. I think this is more clear (i.e. starting from the end in order to process leaves before parents).

---
In reply to: [172637595](https://github.com/dotnet/roslyn/pull/24059#discussion_r172637595) [](ancestors = 172637595)